pyspark/graphar_pyspark/graph.py

```diff
@@ -0,0 +1,215 @@
+limitations under the License.
+"""
+
+from __future__ import annotations
+
+from typing import Optional, Union
+
+from py4j.java_gateway import JavaObject
+from pyspark.sql import DataFrame
+
+from graphar_pyspark import GraphArSession
+from graphar_pyspark.enums import FileType
+from graphar_pyspark.info import GraphInfo
+
+
+class GraphReader:
+    """The helper object for reading graph through the definitions of graph info."""
+
+    @staticmethod
+    def _jvm2py(
+        jvm_result: tuple[
+            dict[str, JavaObject], dict[tuple[str, str, str], dict[str, JavaObject]]
+        ],
+    ) -> tuple[dict[str, DataFrame], dict[tuple[str, str, str], dict[str, DataFrame]]]:
+        """Helper that convert nested Java DataFrames to PySpark DataFrames."""
+        return (
+            {
+                label: DataFrame(jdf, GraphArSession._ss)
+                for label, jdf in jvm_result[0].items()
+            },
+            {
+                complex_label: {
+                    adj_type: DataFrame(jdf, GraphArSession._ss)
+                    for adj_type, jdf in mapping.items()
+                }
+                for complex_label, mapping in jvm_result[1].items()
+            },
+        )
+
+    @staticmethod
+    def read_with_graph_info(
+        graph_info: GraphInfo,
+    ) -> tuple[dict[str, DataFrame], dict[tuple[str, str, str], dict[str, DataFrame]]]:
+        """Reading the graph as vertex and edge DataFrames with the graph info object.
+
+        :param graph_info: The info object for the graph.
+        :returns: Pair of vertex DataFrames and edge DataFrames, the vertex DataFrames are
+        stored as the map of (vertex_label -> DataFrame) the edge DataFrames are stored as a map of
+        ((srcLabel, edgeLabel, dstLabel) -> (adj_list_type_str -> DataFrame))
+        """
+        jvm_result = GraphArSession._graphar.graph.GraphReader.readWithGraphInfo(
+            graph_info.to_scala(), GraphArSession._jss
+        )
+        return GraphReader._jvm2py(jvm_result)
+
+    @staticmethod
+    def read(
+        graph_info_path: str,
+    ) -> tuple[dict[str, DataFrame], dict[tuple[str, str, str], dict[str, DataFrame]]]:
+        """Reading the graph as vertex and edge DataFrames with the graph info yaml file.
+
+        :param graph_info_path: The path of the graph info yaml.
+        :returns: Pair of vertex DataFrames and edge DataFrames, the vertex DataFrames are
+        stored as the map of (vertex_label -> DataFrame) the edge DataFrames are stored as a map of
+        ((srcLabel, edgeLabel, dstLabel) -> (adj_list_type_str -> DataFrame))
+        """
+        graph_info = GraphInfo.load_graph_info(graph_info_path)
+        return GraphReader.read_with_graph_info(graph_info)
+
+
+class GraphWriter:
+    def __init__(self, jvm_obj: JavaObject) -> None:
+        """One should not use this constructor directly, please use `from_scala` or `from_python`."""
```

Comment:
Just a side note: in python, it is almost impossible to prevent users from (not) doing something, and one of the existing patterns is to make it foolproof. It does not seem to be a problem now whether someone creates  `GraphWriter` via `__init__` or via `from_scala`/`from_python`, but in future `__init__` might still be a reasonable way to create a `GraphWriter` object just not very convenient one.

To sum up: it is sort of OK to use `__init__`s, and restricting that via docstrings might be not the most efficient strategy.

Reply:
To be honest, I do not know how else can I restrict users. For example, in docs of `pyspark.sql.DataFrame.__init__` there is a [similar note](https://github.com/apache/spark/blob/master/python/pyspark/sql/dataframe.py#L138):
> A DataFrame should only be created as described above. It should not be directly
    created via using the constructor.
    
I can extend docstrings by adding note and describing the ways how else one can create objects...